src/hooks/common/useSearchKeywordStorage.ts

```diff
@@ -0,0 +1,75 @@
+import { useEffect, useState } from 'react';
+import { LOCAL_STORAGE_KEYS } from 'constants/common';
+
+export function useSearchKeywordStorage() {
+  const [keywords, setKeywords] = useState<string[]>([]);
+  const [isLastKeyword, setIsLastKeyword] = useState<boolean>(false);
+
+  const handleSaveSearchKeyword = (value: string) => {
+    const keywordValue = value.trim();
+    const isValidKeyword = keywordValue.length > 0;
+
+    if (!isValidKeyword) return;
+
+    setKeywords((prevState) => {
+      if (prevState.includes(keywordValue)) {
+        return prevState;
+      }
+
+      if (prevState.length === 8) {
+        return [...prevState.slice(1), keywordValue];
+      }
+
+      return [...prevState, keywordValue];
+    });
+  };
+
+  const handleDeleteSearchKeyword = (value: string) => {
+    setKeywords((prevState) => {
+      const result = prevState.filter((key) => key !== value);
+
+      if (result.length === 0) {
+        setIsLastKeyword(true);
+      }
+
+      return result;
+    });
+  };
+
+  const handleDeleteAllSearchKeyword = () => {
+    localStorage.removeItem(LOCAL_STORAGE_KEYS.searchKeyword);
+    setKeywords([]);
+  };
+
+  useEffect(() => {
+    if (isLastKeyword && keywords.length === 0) {
+      localStorage.removeItem(LOCAL_STORAGE_KEYS.searchKeyword);
+      setIsLastKeyword(false);
+    }
+  }, [isLastKeyword]);
+
+  useEffect(() => {
+    if (keywords.length > 0) {
```

Comment:
note;

**`if (keywords.length > 0)` 조건문이 필요한 이유** 
- 새로고침 시 `const [keywords, setKeywords] = useState<string[]>([]);` 로 인해 로컬스토리지에 저장된 값을 useEffect로 초기화 했으나 다시 빈 배열로 초기화 함
- 저장된 최근 검색어가 모두 삭제되어 해당 조건문을 추가하여 로컬 스토리지에 최근 검색어가 저장되어 있다면 `keywords`가 빈 배열로 초기화 되지 않도록 막음

Reply:
p3;

초기에 로컬 스토리지로부터 keywords를 세팅하는 부분을 useEffect를 쓰지 않고 바로 초기화해주면 코멘트 남겨주신 부분들
1. `if (keywords.length > 0)`를 사용해야 한다.
2. `const [isLastKeyword, setIsLastKeyword] = useState<boolean>(false);` 를 사용해야 한다.
모두 해소될 수 있을 것 같아요!

아래 코드는 러프하게 써본 내용 입니다. 참고 해주세요~
```js
const getValueFromLocalStorage = (key: string): string[] => {
  const localStorageKeyword = localStorage.getItem(key);

  if (localStorageKeyword === '' || localStorageKeyword === null) return [];
  return JSON.parse(localStorageKeyword) as string[];
};

export function useSearchKeywordStorage() {
  // 핵심 코드
  const [keywords, setKeywords] = useState<string[]>(getValueFromLocalStorage(LOCAL_STORAGE_KEYS.searchKeyword)); 
  const [isLastKeyword, setIsLastKeyword] = useState<boolean>(false); // 삭제
 
 ...

// 삭제 
// useEffect(() => {
//   if (isLastKeyword && keywords.length === 0) {
//     localStorage.removeItem(LOCAL_STORAGE_KEYS.searchKeyword);
//     setIsLastKeyword(false);
//   }
// }, [isLastKeyword]);

  useEffect(() => {
    if (keywords.length > 0) { // 삭제 
      localStorage.setItem(
        LOCAL_STORAGE_KEYS.searchKeyword,
        JSON.stringify(keywords),
      );
    }
  }, [keywords]);

// 삭제 
// useEffect(() => {
//   const localStorageKeyword =
//     localStorage.getItem(LOCAL_STORAGE_KEYS.searchKeyword) ?? '[]';

//   if (localStorageKeyword !== 'undefined') {
//     setKeywords(JSON.parse(localStorageKeyword) as string[]);
//   }
// }, []);
...
}
  
```

Reply:
@sohyeonAn 
소현님! 해당 PR을 올리기 전에 소현님 의견과 동일한 방법으로 미리 시도를 해보았는데요.
중요한 부분은 next.js의 특성입니다.

next.js 는 서버사이드 렌더링을 먼저 실행하면서 로컬스토리지 접근 시 `ReferenceError: localStorage is not defined` 에러가 발생합니다.
해당 에러를 해결하기 위해서는 `if(typeof window !== 'undefined')` 조건문 내부에서 로컬스토리지에 접근 또는 `useEffect` 내부에서 접근할 수 있습니다.

제가 작성했던 코드는 다음과 같습니다.
```
  const [keywords, setKeywords] = useState<string[]>(() => {
    if (typeof window !== 'undefined') {
      const savedKeyword = localStorage.getItem(storageKey);
      return savedKeyword ? JSON.parse(savedKeyword) : [];
    }
  }); 
```

위와 같이 작성했을 때 `Hydration failed because the initial UI does not match what was rendered on the server.` 에러가 발생합니다.

<img width="375" alt="image" src="https://github.com/a-daily-diary/ADD.FE/assets/85009583/d0840d03-b1f0-4441-876c-83e33bb59b91">

https://nextjs.org/docs/messages/react-hydration-error

공식문서에는 `useEffect`를 사용하여 해결할 수 있다고 나와있습니다.

실은 다른 next.js 프로젝트에서 `if(typeof window !== 'undefined')` 조건문을 동일하게 적용했을 때 에러가 발생하지 않았는데, 저희 프로젝트에서는 발생해서 당황스럽긴 합니다.
버전의 차이가 있어서 그런건지 잘 모르겠지만..

일단 현재 코드로 적용해 놓되 따로 라이브러리 버전업을 시도해서 동일한 에러가 발생하지 않는다면 소현님의 의견대로 코드를 수정하고자 합니다!

Reply:
음 SSR은 아직 익숙하지 않아서 이런 부분은 생각하지 못했네요 ㅜ
넵 우선 현재 코드로 진행해주세요 🙌